GraphicsContextCG uses memory redundantly for bool fields

kkinnunen-apple opened the pull request:
#### 10d7bc6d7f3c880c1205f0b6c693abdea4537fc9
<pre>
GraphicsContextCG uses memory redundantly for bool fields
<a href="https://bugs.webkit.org/show_bug.cgi?id=267404">https://bugs.webkit.org/show_bug.cgi?id=267404</a>
<a href="https://rdar.apple.com/120837402">rdar://120837402</a>

Reviewed by Antti Koivisto.

Mark some of the held bools as bitfields and reorder.

Before:
Total byte size: 808
Total pad bytes: 61
Padding percentage: 7.55 %

After:
Total byte size: 800
Total pad bytes: 60
Padding percentage: 7.50 %

* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::deferred const): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::consumeHasDrawn):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:

Canonical link: <a href="https://commits.webkit.org/272965@main">https://commits.webkit.org/272965@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa7fa1e7a5a00d8d09df4601e3f9b483e782677d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33593 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12366 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35519 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36210 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30473 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34638 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14731 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9523 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29578 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34068 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10416 "Found 3 new test failures: compositing/updates/animation-non-composited.html, fast/forms/state-restore-to-non-edited-controls.html, scrollingcoordinator/scrolling-tree/fixed-inside-frame.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29949 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9106 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9190 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29956 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37535 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30484 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30283 "Found 1 new test failure: fast/forms/state-restore-to-non-edited-controls.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35332 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9328 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7284 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33214 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11106 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7789 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9912 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10098 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
```diff
@@ -152,12 +152,12 @@ class WEBCORE_EXPORT GraphicsContextCG : public GraphicsContext {
-    // Flag for pending draws. Start with true because we do not know what commands have been scheduled to the context.
-    bool m_hasDrawn { true };
     mutable std::optional<DestinationColorSpace> m_colorSpace;
+    const RenderingMode m_renderingMode : 1; // NOLINT
+    const bool m_isLayerCGContext : 1;
+    mutable bool m_userToDeviceTransformKnownToBeIdentity : 1 { false };
+    // Flag for pending draws. Start with true because we do not know what commands have been scheduled to the context.
+    bool m_hasDrawn : 1 { true };
```
Bit of an overkill perhaps for a type that won't have many instances alive? Isn't this a stack allocated type anyway?

kkinnunen-apple replied:
No, heap allocated.
If we have 20000 layers or 2d contexts, we use 160 mb of GPUP's 300mb limit..

kkinnunen-apple committed:
GraphicsContextCG uses memory redundantly for bool fields
https://bugs.webkit.org/show_bug.cgi?id=267404
rdar://120837402

Reviewed by Antti Koivisto.

Mark some of the held bools as bitfields and reorder.

Before:
Total byte size: 808
Total pad bytes: 61
Padding percentage: 7.55 %

After:
Total byte size: 800
Total pad bytes: 60
Padding percentage: 7.50 %

* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::deferred const): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::consumeHasDrawn):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:

Canonical link: https://commits.webkit.org/272965@main
